ens.asciidoc

```diff
@@ -13,31 +18,75 @@ The _Ethereum Name Service (ENS)_ is one of the first Ethereum DApps, offering a
 
 === History of Ethereum name services
 
-ENS is not the first name registration service in Ethereum. It was preceded by "NameReg" ...
+Name registrations is the first non-currency application of blockchains, being pioneered by Namecoin. The ethereum whitepaper contains as one of their example aplications, a two lines "namecoin-like" registration system. The early releases of geth and CPP client had a standard default "namereg" contract (not used anymore) and many proposals and ERCs have existed since the beginning, but it was not before Nick Johnson started working for Ethereum Foundation in 2016 and took the project under his wing that work on a serious registrar started.
+
+ENS was launched on Star Wars day, May 4, 2017 (after a failed attempt of launching it on Pie Day, march 15) and it was open for registrations.
+
+=== The ENS specification
+
+ENS is specified mainly in three Ethereum Improvement Proposals (EIP), namely ERC137 which specifies the basic ENS functioning, ERC162 the auction system for the .eth root and ERC181 for reverse registration of addresses.
+
+ENS folows a "sandwich" design philosophy: a very simple layer on the bottom, followed by layers of more complex but replaceable code, followed with a very simple top layer that keeps all the funds in separate individual accounts.
+
+=== Bottom of the sandwich: Name Owners and resolvers
+
+The base layer of the sandwich is a cleverly simple (less than 50 lines of code) contract defined by ERC137 that basically only allows owners of names to set information about their names and to create subnodes (subdomains). The core of the simplicity is the identifier mechanism for names, called namehash.
+
+Namehash is an algorithm for converting any name in a identifier hash of the name, and it works like this:
+
+Root Node:  `keccak(0x0000000000000000000000000000000000000000000000000000000000000000)`
+".eth" node: `keccak( (Root Node) + keccak('eth'))`
+"example.eth" node: `keccak( ('.eth' node) + keccak('example'))`
+"subdomain.example.eth" node: `keccak( ('example.eth' node) + keccak('subdomain'))`
+
+And recursively so. The only functions on the base layer are those that enable a node owner to set information about their own node (specifically: resolver, time to live and transferring ownership) and to create an owner to a subnode.
+
+==== Rootnode ownership
+
+One of the results of this hierarchical system is that it relies on the owners of the Rootnode, that are able to create top level domains. While the eventual goal is to move this into some sort of decentralized decision making process for new top level domains, right now the rootnode is a 4 out of 7 multisig, held by people in different countries (built as a reflection of the seven keyholders of the DNS system). Currently the purpose and goal of these keyholders is to work in consensus with the community to:
```

Comment:
For even greater clarity, I would add;
"A majority of at least 4 of the 7 keyholders is required to effect any change."